### ISSUE TITLE
Reproducible builds

### ISSUE THREAD

**Issue by rugk** [stefan-niedermann/nextcloud-notes#1264]
Please use **[GitHub reactions](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/)** 👍 to show that you are affected by the same issue. Please don't comment if you have no relevant information to add!

**Is your feature request related to a problem? Please describe.**

Faster updates from a trusted F-Droid repository.
And reporducible builds are alos very good for security reasons: See https://reproducible-builds.org/.

>i unfortunately can't provide a quicker way to test new approaches than the release cycle of F-Droid allows

https://github.com/stefan-niedermann/nextcloud-notes/issues/1256#issuecomment-861489993

Technically you could if the builds are reproducible you can sign both the same APK with your key and F-Droid just distributes that.

**Describe the solution you'd like**

F-Droid's [reproducible builds](https://f-droid.org/en/docs/Reproducible_Builds/)).

**Describe alternatives you've considered**

**Additional context**

The people at CCTG and Signal (IIRC) e.g. made it work, so you can built on their experience: https://codeberg.org/corona-contact-tracing-germany/cwa-android/src/branch/main/docs/rebuilding.md

It's basically a Docker image with F-Droids buildserver.

Note: Yes, I know it's likely not easy to setup, but it is a good idea for the future. :blush:

**Comment by rugk** [comment]
> The build time itself is not the issue, but the time of the index update. Therefore this will not speed up the rollout of the app on F-Droid.

No, AFAIK, one could install the app from F-Droid, and then switch to the installed app from your source (if it is no F-Droid repo, even a simple APK attached at GitHub releases would be sufficient e.g.). The advantage then is testers could just get your APK before it is built by F-Droid, as it is signed with the same public key.
I know I did this for CCTG, so I know it technically works. :slightly_smiling_face:

**Comment by stefan-niedermann** [comment]
I don't fully understand - even if F-Droid would accept a build of mine as equal and ship it, Android would prevent one from installing it over an already installed version due to a signing key mismatch...?

> as it is signed with the same key.

In fact it *isn't* the same key though, is it? - F-Droid signs with its own key, while apk files build by me are signed with my key, even if the builds are the same...

**Comment by rugk** [comment]
You sign it with your own key, F-Droid builds the same binary and verifies it is identical and if so,. distributes your binary.
That is one way, AFAIK. That’s the whole clever idea of that. :slightly_smiling_face: 

I hightly suggest reading https://f-droid.org/en/docs/Reproducible_Builds/, especially the “How it is implemented as of now” part.

**Comment by stefan-niedermann** [comment]
So what exactly needs to be done on our side? As far as i understand, we would need to setup a production build pipeline (currently only dev build) which would require us to place the signing key on as a variable on the CI runner (GitHub in this case).
Since i don't trust GitHub i am not willing to push my private key as a secret variable there. Neither do i have the capabilities to setup and maintain an own build server.

Am i missing something? Don't get me wrong, reproducible builds are a fine thing, though i can't see how we should solve this predicament...

**Comment by stefan-niedermann** [comment]
Well i read the docs but i need more exact instructions regarding the steps to do. "Build your APK locally and upload it anywhere the F-Droid server can find it" is unfortunately not preceise enough for me to get this done...

**Comment by IzzySoft** [comment]
Ah, that part I can answer. Let me cut it into pieces:

* "Build your APK locally": ideally using the same tool chain (fdroidserver) so the resulting APK is binary identical to what F-Droid builds – apart from the signature (you will need to sign this APK with **your** key)
* "upload it anywhere the F-Droid server can find it": it must be publicly available, i.e. F-Droid must be able to access it
* the piece you're probably missing: that URL then needs to be added to the build block in F-Droid's `*.yml`. And it should carry the `versionCode` in it's name, so it's easily matched – like `org.example_<versionCode>.apk` which then would be referenced as `org.example_%c.apk` and the `%c` would then always be replaced by the `versionCode` of the current build.

Does that help?
